### PR TITLE
Font Fallback Support (Linux only)

### PR DESCRIPTION
**Pull request by hlieberman**
This obsoletes and replaces #441.

### CODE REVIEW

File: font/src/ft/list_fonts.rs [hunk 1]
```diff
@@ -353,6 +373,23 @@ pub mod fc {
...
         }
 
+
+        /// Add charset to the pattern
+        ///
+        /// The referenced charset is copied by fontconfig internally using
+        /// FcValueSave so that no references to application provided memory are
+        /// retained. That is, the CharSet can be safely dropped immediately
+        /// after being added to the pattern.
+        pub unsafe fn add_charset(&self, charset: &CharSetRef) -> bool {
```
Comment (jwilm) [inline font/src/ft/list_fonts.rs]: Oops, we can remove the `unsafe` here. My bad.

**Comment by hlieberman** [comment]
This is ready for your review again, @jwilm.

File: font/src/ft/mod.rs [hunk 2]
```diff
@@ -209,8 +164,106 @@ impl FreeTypeRasterizer {
...
+        let mut charset = fc::CharSet::new();
+        charset.add(glyph);
+        let mut pattern = fc::Pattern::new();
+        pattern.add_charset(&charset);
+
+        let config = fc::Config::get_current();
+        match fc::font_match(config, &mut pattern) {
+            Some(font) => {
+                if let (Some(path), Some(index)) = (font.file(0), font.index(0)) {
+                    match self.keys.get(&path.to_path_buf()) {
```
Comment (jwilm) [inline font/src/ft/mod.rs]: Can just pass `&path` here.

File: font/src/ft/mod.rs [hunk 3]
```diff
@@ -209,8 +164,106 @@ impl FreeTypeRasterizer {
...
+                            let face = self.library.new_face(path, index)?;
+                            let key = FontKey::next();
+                            self.keys.insert(pathbuf, key);
+                            self.faces.insert(key, face);
+                            return Ok(key)
+                        }
+                    }
+                }
+
+                Err(Error::MissingFont(
```
Comment (jwilm) [inline font/src/ft/mod.rs]: If we move this into an `else` for the preceding `if let`, we can omit `return` on the `return Ok(key)` lines above.

File: font/src/ft/mod.rs [hunk 4]
```diff
@@ -209,8 +164,106 @@ impl FreeTypeRasterizer {
...
+                        // We've previously loaded this font, so don't
+                        // load it again.
+                        Some(&key) => {
+                            debug!("Hit for font {:?}", path);
+                            return Ok(key)
+                        },
+
+                        None => {
+                            debug!("Miss for font {:?}", path);
+                            let pathbuf = path.to_path_buf();
```
Comment (jwilm) [inline font/src/ft/mod.rs]: This clone isn't necessary either.

**Pull request merged**